components/style/values/specified/length.rs

```diff
@@ -222,14 +222,96 @@ impl CharacterWidth {
     }
 }
 
+const ABSOLUTE_LENGTH_MAX: f64 = (1 << 30) as f64;
+const ABSOLUTE_LENGTH_MIN: f64 = - (1 << 30) as f64;
+
+macro_rules! impl_from_au {
+    ($ty: ty, $multiplier: expr) => {
+        impl From<$ty> for Au {
+            fn from(length: $ty) -> Au {
+                Au(
+                    ((length.0 as f64) * ($multiplier as f64))
+                    .max(ABSOLUTE_LENGTH_MIN)
+                    .min(ABSOLUTE_LENGTH_MAX)
+                    .round() as i32
+                )
+            }
+        }
+    }
+}
+
+/// A wrapper representing a floating-point pixel (px) length
+#[derive(Clone, PartialEq, Copy, Debug, HeapSizeOf)]
+pub struct Px(pub f32);
+
+impl_from_au!(Px, AU_PER_PX);
+
+/// A wrapper representing a floating-point inches (in) length
+#[derive(Clone, PartialEq, Copy, Debug, HeapSizeOf)]
+pub struct In(pub f32);
+
+impl_from_au!(In, AU_PER_IN);
+
+/// A wrapper representing a floating-point centimeter (cm) length
+#[derive(Clone, PartialEq, Copy, Debug, HeapSizeOf)]
+pub struct Cm(pub f32);
+
+impl_from_au!(Cm, AU_PER_CM);
+
+/// A wrapper representing a floating-point millimeter (mm) length
+#[derive(Clone, PartialEq, Copy, Debug, HeapSizeOf)]
+pub struct Mm(pub f32);
+
+impl_from_au!(Mm, AU_PER_MM);
+
+/// A wrapper representing a floating-point quarter-millimeter (q) length
+#[derive(Clone, PartialEq, Copy, Debug, HeapSizeOf)]
+pub struct Q(pub f32);
+
+impl_from_au!(Q, AU_PER_Q);
+
+/// A wrapper representing a floating-point point (pt) length
+#[derive(Clone, PartialEq, Copy, Debug, HeapSizeOf)]
+pub struct Pt(pub f32);
```

Comment:
Let's use `CSSFloat` here. They're aliases now, but worth to be consistent.

Also, probably implementing `ToComputedValue` would be a bit better than `From<Au>`, but not necessarily worth I guess.